File: crud-item-list/src/main/java/com/gurunars/crud_item_list/ActionMoveUp.java
package com.gurunars.crud_item_list;

import java.util.List;
import java.util.Set;

import java8.util.function.BiFunction;
import java8.util.function.Function;

final class ActionMoveUp<ItemType> implements Action<ItemType> {

    private final BiFunction<List<ItemType>, Set<ItemType>, List<Integer>> positionFetcher =
            new PositionFetcher<>();

    private final Function<List<Integer>, Boolean> solidChunkChecker =
            new CheckerSolidChunk();

    @Override
    public void perform(List<ItemType> all, Set<ItemType> selectedItems) {
        List<Integer> positions = positionFetcher.apply(all, selectedItems);
        int positionToMoveDown = positions.get(0) - 1;
        ItemType itemToMoveDown = all.get(positionToMoveDown);
        all.add(positions.get(positions.size() - 1) +1, itemToMoveDown);
        all.remove(positionToMoveDown);
    }

    @Override
    public boolean canPerform(List<ItemType> all, Set<ItemType> selectedItems) {
        List<Integer> positions = positionFetcher.apply(all, selectedItems);
        return solidChunkChecker.apply(positions) && positions.get(0) > 0;
    }

}

File: crud-item-list/src/main/kotlin/com/gurunars/crud_item_list/ActionMoveUp.kt
package com.gurunars.crud_item_list

import com.gurunars.box.IRoBox
import com.gurunars.box.merge
import com.gurunars.box.oneWayBranch
import com.gurunars.item_list.Item

internal class ActionMoveUp<ItemType : Item> : Action<ItemType> {

    override fun canPerform(
        all: IRoBox<List<ItemType>>,
        selectedItems: IRoBox<Set<ItemType>>
    ): IRoBox<Boolean> = merge(all, selectedItems).oneWayBranch {
        val positions = getPositions(first, second)
        isSolidChunk(positions) && positions[0] > 0
    }

    override fun perform(
        all: List<ItemType>,
        selectedItems: Set<ItemType>,
        consumer: ItemSetChange<ItemType>
    ) {
        val positions = getPositions(all, selectedItems)
        val positionToMoveDown = positions[0] - 1
        val itemToMoveDown = all[positionToMoveDown]
        consumer(mutableListOf<ItemType>().apply {
            addAll(all)
            add(positions[positions.size - 1] + 1, itemToMoveDown)
            removeAt(positionToMoveDown)
        }, selectedItems)
    }

}
